homeassistant/components/vicare/water_heater.py

```diff
@@ -153,3 +157,9 @@ def current_operation(self):
     def operation_list(self):
         """Return the list of available operation modes."""
         return list(HA_TO_VICARE_HVAC_DHW)
+
+    def set_operation_mode(self, operations_mode):
```

Comment:
If one time charge is not an operation mode reflected as current operation we should not use the set operation mode service to activate this feature. Use a custom entity service instead.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
I'm not sure. In the end one time charge should be reflected also in the state.
Unfortunately there is no data point on the api which tells us if it is still active. Usually it's just a matter of minutes, so I guess the implementation here is not so far-off. The one-time-charge state will be shown as active until the next update of the entity.

Reply:
If the api doesn't return it as a current operation state we shouldn't do that either. We should just reflect the device state not invent our own.

Reply:
@MartinHjelmare as @oischinger rightfully point out the API from Viessmann is not supporting a rich enough API to actually show the state changes. This does not mean that the device does not have that state though (so in other words it is not inventing a state - the state is shown on the heater however not in the API), so this is bridging between lack of Viessmann API and reality of the Viessmann devices. That on the other hand should not mean that with adequate effort it should not be possible to utilize this important feature from Home Assistant, done the right way this actually helps to save a lot of energy for hot water heating.

If I were to decide on which principle wins here I would suggest Home Assistant supports helping users save energy (in this case often oil or natural gas) rather than having to wait for the API to be "correct", and hence do this completely correct. I would hence recommend the change as it serves a purpose. I understand that creating custom entity services is a way forward however it is also a larger effort to create. That was just my IMHO :-)

Reply:
Sorry, but we don't allow creating state in device or service integrating integrations that the 3rd party API doesn't support.